src/contracts/AaveCLRobotOperator.sol

```diff
@@ -0,0 +1,166 @@
+import {IKeeperRegistrar} from '../interfaces/IKeeperRegistrar.sol';
+import {IKeeperRegistry} from '../interfaces/IKeeperRegistry.sol';
+
+/**
+ * @title AaveCLRobotOperator
+ * @author BGD Labs
+ * @dev Operator contract to perform admin actions on the automation keepers.
+ *      The contract can register keepers, cancel it, withdraw excess link,
+ *      configure the registered keepers and disable automation on a certain proposal.
+ */
+contract AaveCLRobotOperator is IAaveCLRobotOperator {
+  /// @inheritdoc IAaveCLRobotOperator
+  address public immutable LINK_TOKEN;
+
+  address internal _fundsAdmin;
+  address internal _maintenanceAdmin;
+  address internal _linkWithdrawAddress;
+
+  mapping(address upkeep => KeeperInfo) internal _keepers;
+
+  mapping(address upkeep => mapping(uint256 proposalId => bool isDisabled))
+    internal _disabledProposals;
+
+  /**
+   * @dev Only funds admin can call functions marked by this modifier.
+   */
+  modifier onlyFundsAdmin() {
+    require(msg.sender == _fundsAdmin, 'CALLER_NOT_FUNDS_ADMIN');
+    _;
+  }
+
+  /**
+   * @dev Only maintenance admin or funds admin can call functions marked by this modifier.
+   */
+  modifier onlyMaintenanceOrFundsAdmin() {
+    require(
+      msg.sender == _maintenanceAdmin || msg.sender == _fundsAdmin,
+      'CALLER_NOT_MAINTENANCE_OR_FUNDS_ADMIN'
+    );
+    _;
+  }
+
+  /**
+   * @param linkTokenAddress address of the ERC-677 link token contract.
+   * @param linkWithdrawAddress withdrawal address to send the exccess link after cancelling the keeper.
+   * @param fundsAdmin address of funds admin.
+   * @param maintenanceAdmin address of the maintenance admin.
+   */
+  constructor(
+    address linkTokenAddress,
+    address linkWithdrawAddress,
+    address fundsAdmin,
+    address maintenanceAdmin
+  ) {
+    LINK_TOKEN = linkTokenAddress;
+    _linkWithdrawAddress = linkWithdrawAddress;
+    _fundsAdmin = fundsAdmin;
+    _maintenanceAdmin = maintenanceAdmin;
+  }
+
+  /// @inheritdoc IAaveCLRobotOperator
+  function register(
+    string memory name,
+    address upkeepContract,
+    uint32 gasLimit,
+    bytes memory checkData,
+    uint96 amountToFund,
+    address keeperRegistry,
+    address keeperRegistrar
+  ) external onlyFundsAdmin returns (uint256) {
+    (IKeeperRegistry.State memory state, , ) = IKeeperRegistry(keeperRegistry).getState();
+    uint256 oldNonce = state.nonce;
+
+    bytes memory payload = abi.encode(
+      name,
+      0x0,
+      upkeepContract,
+      gasLimit,
+      address(this),
+      checkData,
+      amountToFund,
+      0,
+      address(this)
+    );
+    bytes4 registerSig = IKeeperRegistrar.register.selector;
+    LinkTokenInterface(LINK_TOKEN).transferAndCall(
+      keeperRegistrar,
+      amountToFund,
+      bytes.concat(registerSig, payload)
+    );
+
+    (state, , ) = IKeeperRegistry(keeperRegistry).getState();
+    if (state.nonce == oldNonce + 1) {
```

Comment:
What exactly is this condition for? I assume the auto-approve, but better to document a bit what is it about

Reply:
Yes the nonce increments when keeper has been registered and is auto-approved, have added some comments